data/service/api/v1/alerts.go

```diff
@@ -31,23 +30,25 @@ func DeleteAlert(dCtx service.Context) {
 	details := request.DetailsFromContext(ctx)
 	repo := dCtx.AlertsRepository()
 
-	if err := checkAuthentication(details, r.PathParam("userID")); err != nil {
```

Comment:
Given that all three of these endpoints have the same path parameters (`userID` and `followedID`) could we DRY this up? Perhaps in a common authorization function? This might simplify some of the extra functions at the end of this file into one "validateAndAuthorize" function that might return the `userID`, `followedID`, and an error.

Reply:
I've studied this a bit, and even created the function in question. What I came up with, looked like this:

```go
// checkAlertsPrerequisites handles HTTP administration tasks.
//
// Such as authentication, authorization, and consistency checks.
func checkAlertsPrerequisites(dCtx service.Context) (string, string, *platform.Error) {
	r := dCtx.Request()
	ctx := r.Context()
	details := request.DetailsFromContext(ctx)

	if err := checkAuthentication(details); err != nil {
		return "", "", platform.ErrorUnauthorized()
	}

	followedID := r.PathParam("followedID")
	userID := userIDWithServiceFallback(details, r.PathParam("userID"))
	pc := dCtx.PermissionClient()
	if err := checkUserAuthorization(ctx, pc, userID, followedID); err != nil {
		return "", "", platform.ErrorUnauthorized()
	}

	if err := checkUserIDConsistency(details, r); err != nil {
		return "", "", platform.ErrorUnauthorized()
	}

	return userID, followedID, nil
}
```
This function is responsible for a lot of things. Namely:

1. authenticating the request
2. authorizing the user
3. checking the consistency of the userIDs

To do that, it has knowledge of:

1. the data package's `Context`
2. the rest package's `Request`
3. the platform package's `Context`
4. the platform package's `Error`
5. the request's path structure

That seems like a lot for what's supposed to be a helper function. It would also be suspect to increased complexity if a future handler with slightly different requirements were to be added.[^1]

I would prefer to keep things simpler, even if it means repeating myself. 

The three existing helpers (`checkUserAuthorization`, `checkAuthentication`, and `checkUserIDConsistency`) know only a small part of all of that which goes into a handler function. As a result, they're easier to read and reason about. They're easier to write tests that exercise them as well, as there's less setup required, and fewer internal logical paths to trace.

P.S. It's true that the handler functions (`GetAlert`, `DeleteAlert`, and `UpsertAlert`) themselves know about all of the things that `checkAlertsPrerequisites` would, but they have no choice in the matter, that's complexity inherent to how HTTP handlers in the platform package are written. If I can contain/limit that complexity to just those handlers, keeping the helper functions smaller, simpler, and focused on a single responsibility, then that strikes me as a good choice.

1. The most straightforward path would be to modify the `checkAlertsPrerequisites` function to handle the new needs, but that would make an already complex function more so, and perhaps even worse, it could easily have unforeseen effects on the behavior of the other handler functions. Adding a separate `checkAlertsPrequisites2` of some sort, just makes things more complex as well.

Reply:
Okay, I see your point. That does make it more complex. Another option you could consider is adding much of the code as a middleware function, just like the `Authenticate` middleware used in `AlertRoutes`.

One of the reasons I mentioned this was that I noticed that the "authentication/authorization/validation" code is different for each of `DeleteAlert`, `GetAlert`, and `UpsertAlert`, but it seems like it should be all the same. For example, in `GetAlert` the check for `checkUserIDConsistency` is after the `checkUserAuthorization` (which we really want last since it is the most expensive operation). Also, the `checkUserAuthorization` is after `DecodeRequestBody` in `UpsertAlert` (though one could argue that validating the request body should go before all authentication/authorization since it is validating the incoming request before trying to do anything with it).

Reply:
The middleware idea is tempting, but I prefer to have middleware deal with the concerns of HTTP (or whatever transport layer). Keeping the transport layers and business logic layers separate yields benefits in terms of clearer logic and easier testing. 

The orderings of the helpers do differ a little, and most of those differences were guesses at an performance optimization that I've never measured, but do believe with probability 0.99 don't matter anyway :). So I think they could all be done in the same order and it's very likely that no one would ever know the difference.

> Also, the checkUserAuthorization is after DecodeRequestBody in UpsertAlert (though one could argue that validating the request body should go before all authentication/authorization since it is validating the incoming request before trying to do anything with it).

I wouldn't necessarily agree with that. Though the argument is largely academic, as a successful request needs to perform all checks and operations anyway, so the order doesn't matter.

But to continue with the thought experiment... I would generally want to perform the operations in shortest time to completion order, as that lets me abort at the soonest opportunity, thereby optimizing for minimal resource consumption. I generally think of json un/marshaling as being relatively slow, due to the use of reflect and the amount of object allocation involved.

In this case, the permission check is slower than typical, as it requires a round trip to the gateway service. I don't have a good feel for how it compares to json un/marshaling.

I'd also consider that in the event that a request fails the authentication or authorization steps, it increases the likelihood that the request body is either malformed (or even malicious). So leaving the expensive json un/marshaling operation for last seems like a good idea.

Reply:
While I don't think a code change is required, I'm going to have to disagree on "Though the argument is largely academic, as a successful request needs to perform all checks and operations anyway, so the order doesn't matter."

In most cases, I validate input *before* doing anything else. There are a couple of reasons for this:
- If the input is invalid, why bother doing anything else or invoking any other microservices or the database? (If a bad actor is trying to hack the API, then catching invalid input before impacting another microservice/database is a good thing. Think of a DDOS attack with bad input - would you rather have just this microservice impacting by the attack or this microservice and the permissions microservice and/or the database?)
- Unexpected consequences can occur if any input is used prior to its validation.

With both of the above, I try to validate all input first as a matter of habit.

Either way, its up to you.